src/fmu/dataio/dataio.py

```diff
@@ -420,6 +420,7 @@ class ExportData:
     _rootpath: Path = field(default_factory=Path, init=False)
 
     def __post_init__(self) -> None:
+        assert isinstance(self.config, dict)
```

Comment:
Yes, it will exit for the user. We use these asserts once in a while for mypy purposes to allow mypy to be confident something is a particular type. In general, we try to avoid it... but sometimes it is the only way to solve ambiguity.

In this case we have

```python
config: dict | GlobalConfiguration = field(default_factory=dict)
```

At initialization `config` will default to a dict, but eventually we want it to be a `GlobalConfiguration`.

```python
def __post_init__(self) -> None:
    assert isinstance(self.config, dict)
```

Here, we can be sure that `self.config` is a dict because its default initialization is such, and this method is called right after that. However, mypy is not sure. It just knows that `config` can be either a `dict` or a `GlobalConfiguration` and this ambiguity on mypy's end can cause mypy to say we're doing wrong things later on. But we know that we are not. So asserting that config is a dict let's mypy share our understanding and if we are correct then we know this assert won't cause an exit, it will always be true.

In my view this is usually a code smell, but sometimes it simply cannot be avoided because mypy isn't yet smart enough to intuit this. It happens most frequently with unioned types, and using dataclasses can create patterns where this is unavoidable.

Reply:
Ah, I see. Makes sense to use it if mypy is being difficult. Thanks for the explanation!